_posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md

```diff
@@ -0,0 +1,79 @@
+
+One of the most popular, battle-tested and performant asynchronous frameworks is [Sidekiq](https://sidekiq.org), which is multi-threaded and utilises Redis for it's queuing storage. It’s an open source library with two paid tiers for additional high-end features.
+
+Disclaimer: I'm a massive fan and “Pro” customer of Sidekiq and have met [Mike](https://twitter.com/getajobmike) (it's author) a little at a couple of Ruby conferences. That hasn't coloured my view of the library, it has _always_ been terrific and is worth every penny for our use at [work](https://coveragebook.com).
+
+
+## Instead of…
+
+…defining your jobs using Active Job:
+
+```ruby
+class DoThingsInBackgroundJob < ApplicationJob
+  queue_as :default
+
+  def peform(an_active_record_object)
+    an_active_record_object.do_things
+  end
+end
+```
+
+## Use…
+
+…Sidekiq directly:
+
+```ruby
+class DoThingsInBackgroundJob
+  include Sidekiq::Worker
+  sidekiq_options queue: "default"
+
+  def peform(id)
+    an_active_record_object = ActiveRecordObject.find_by(id: id)
+    an_active_record_object.do_things
+  end
+end
+```
+
+
+## Why?
+
+Active Job is a solid super-set of Sidekiq's functionality, but you’re missing out on a number of advantages if you use it without.
+
+If you're doing _a lot_ of work in the background, enqueuing many small fast-running jobs (as is generally suggested) there is a [2—20x performance improvement](https://github.com/mperham/sidekiq/wiki/Active-Job#performance) if you use Sidekiq directly. This improvement will be dependant on your specific setup.
+
+If you are enqueing lots of jobs, you could be using [Sidekiq’s bulk enqueing functionality](https://github.com/mperham/sidekiq/wiki/Bulk-Queueing), this is hard to do performantly when using sidekiq with ActiveJob.
+
+When using Sidekiq via Active Job there are, confusingly, multiple levels of retries in the event of a job failure. Active Job has it's own retry mechanism, which once completed then passes down to Sidekiq’s own (seperate and different!) retry system. This is a strange thing to debug when things go wrong.
+
+When using Active Job and passing in an Active Record object, you rely on the object being serialized to a text argument using [Global ID](https://github.com/rails/globalid). This does save doing the lookup yourself (as you can see in the Active Job example) but it raises errors if records can be deleted (or even not created!) before the job is pulled form the queue. It also makes the arguments of the jobs a bit ugly in the web dashboard.
+
+There’s no real harm to “locking yourself into” a specific job framework. You are unlikely to swap in future for another queuing system with similar (intentionally) limited functionality availible in Active Job. In a real life situation where your asyncronous infrastructure is struggling, you’d be assessing various solutions with more functionality and planning for a major migration project.
```

Comment:
I think quotes should only be around "locking"

*You are unlikely to switch to another queuing system...

End of this paragraph feels messy/out of place.